x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -150,8 +151,14 @@ private boolean shouldKeepToken(
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
         final MappedFieldType ft = context.getFieldType(fieldName);
         if (ft == null) {
-            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+            throw new ElasticsearchParseException("[" + fieldName + "]" + " is not a mapped field");
         }
+
+        final String fieldTypeName = ft.typeName();
+        if (AllowedFieldType.isTypeNameValid(fieldTypeName) == false) {
+            throw new ElasticsearchParseException("[" + fieldTypeName + "]" + " is not an appropriate field type for text expansion query");
```

Comment:
It would probably be a friendlier error to also specify the allowed field types.

Reply:
Updated.